stmt.go

```diff
@@ -40,8 +40,8 @@ import (
 
 type stmtOptions struct {
 	boolString         boolString
-	fetchArraySize     int
-	prefetchCount      int
+	fetchArraySize     int // zero means DefaultFetchArraySize negative means 0
+	prefetchCount      int // zero means DefaultPrefetchCount
```

Comment:
And what does negative mean?  

Overall the definitions do seem a little complex, as @anthony-tuininga noted.  I know you said you were trying to be consistent with other values.  Maybe @tgulacsi has some comments.

Reply:
The behaviour of these attributes is kept similar to the way we handle MaxIdleConnections.

Reply:
I think zero array size won't work, so we can leave 0 as default, any other as is.

Reply:
fetchArraySize cannot be 0.
prefetchCount can be 0, that was the default before ODPI v4.

Reply:
ODPI would actually have used prefetch of 2 since that was hardcoded. The ODPI v4 default prefetch is still 2.  But now in ODPI 4 the prefetch can be set by the user to >= 0.  How do you want this represented in godror?

Reply:
I will try to clarify the comments with few details . Please see

Go initializes all variables inside struct to 0. so if user doesnt overwrite these parameters, fetchArraySize and prefetchCount will be zero. in the comment, when 0 is mentioned, it points to Go initialization. By default they have 0, but when we sent to OCI, they are sent as DefaultPrefetchCount and DefaultArraySize values respectively. 

Now if user overwrites it to 0, then internally we modify it to -1 to differentiate from Go initialization. So when comment says -1, it  indicates that we are going to pass 0 to OCI. when user passes 0, it is passed as 0 to OCI . In case of ArraySize, it returns error. Now if user sets -1 or any negative value, we set internally as -1 and pass 0 to OCI. 

I went through code for returning error, when user sets -1 , it looks not trivial change. @tgulacsi  , can you correct?

Reply:
If the user does not set anything, then stmtOptions.fetchArraySize will be 0. 
That should mean DefaultArraySize.

For prefetchCount, 0 means uninitialized, which should also mean the default, DefaultPrefetchCount.
In the PrefetchCount option, we can store -1, which can mean 0.

What if we store prefetchCount+1, and subtract 1 in stmtOption.PrefetchCount() ?
then 0 is the uninitialized -> DefaultPrefetchCount,
and anything else is one more than what has been set through the PrefetchCount option.

Reply:
If i understand correctly, above will allow DefaultPrefetchCount to be sent if user hasn't set anything. If he sets 0, it passes 0 to OCI. If he sets -1/-2, do we plan to throw error or treat is as 0 or defaults like DefaultPrefetchCount .

should we throw error for -1/-2,.. values from go driver?

current patch achieves the above requirement, it treats -1/-2 as 0. no error is thrown. however  error is thrown for 0 ArraySize from ODPI. For PrefetchCount, 0 is valid.

Reply:
The truth is that we can use ANYTHING in stmtOptions - as it is only set through the godror.PretechCount option (function), that can encode the user supplied value  in any way.
This encoding is what stmtOption.PrefetchCount should interpret and return the ready-to-use value.

For user setting, negative values are uninterpretable, so use the default.
Zero should mean zero.
Then we can encode it as "-1 means zero", "zero means default", "positive value means itself".
Or as I suggested, "zero means default", "positive means one more than the real value".
Both are acceptable, I favor mine :)

I just want to keep the 0 mean "not set by the user, so use the default", to let the uninitialized stmtOption be usable as is.

TL;DR; 
  - godror.PrefetchCount option should accept anything, <=0 means Default, >0 means  that number;
  - encode it as you feel in stmtOption.prefetchCount, but 0 must mean default;
  - stmtOption.PrefetchCount must decode this and return the number to be passed on to ODPI.

Reply:
when you say, <=0 means Default,>0 means that number. how does user pass 0. 
0 is special right? if user sets 0, he meant to disable prefetch which cant be DefaultPrefetchCount. if he doesnt set anything, we take default, DefaultPrefetchCount.

1. applying your logic, user sets negative, it goes to default. if he sets 0, it gets disabled. else number passed as is
user doesnt set anything, we get -1 (0-1) in stmtOption.prefetchCount. we pass DefaultPrefetchCount down
user sets -1, we get -1 (0-1) in stmtOption.prefetchCount. we pass DefaultPrefetchCount down
user sets 0, we get 0 in stmtOption.prefetchCount. we pass 0 down

2. I was doing this:
if user sets 0 or -1, disable PreFetchCount. else number. passed as is


you see option 2 not good :)

Reply:
You seem to miss the fact that stmtOption.prefetchCount is an encoded value, which is set by godror.PrefetchCount, read by stmtOption.PrefetchCount - both controlled by us.
So there's no "user sets ...", only "user gives n to godror.PrefetchCount, which stores m, which is decoded to k by stmtOption.PrefetchCount, which is used as-is".